kbs/docs/initdata.md

```diff
@@ -0,0 +1,267 @@
+
+There are multiple ways to binding the integrity of initdata to the TEE evidence.
+Many TEE platforms supports HOSTDATA field. The HOSTDATA field could be set by
+the untrusted host when launching the TEE, and the field will be included in the
+TEE evidence for remote attestation.
+
+Platforms and corresponding field of the evidence
+- Intel TDX: `mr_config_id`, 48 bytes. Actually `mr_owner` and `mr_owner_config` have similiar
+attributes, but we select only `mr_config_id` for such use.
+- AMD SNP: `hostdata`, 32 bytes.
+- Arm CCA: `CCA_REALM_PERSONALIZATION_VALUE`, 64 bytes.
+- Intel SGX: `CONFIGID`, 64 bytes.
+
+When users want to deploy a TEE, they need to prepare an initdata metadata. The host
+(probably untrusted) SHOULD start TEE instance with metadata's `digest` as hostdata
+(or calculate `digest` with `algorithm` and `data`).
+
+The software outside the TEE would deliver the initdata metadata into the TEE in
+some way due to concrete architecture design. This spec does not define the concrete
+way to deliver initdata metadata, but strongly recommend that before using the received
+initdata, software stack inside the TEE MUST check whether the digest of the received
+initdata is aligned with the one inside evidence.
+
+Other platforms, such as (v)TPM based platforms, can record the initdata digest
+by extending the PCR before using it. This way will also accomplish the integrity binding
+to the TEE evidence.
+
+### Data Canonicalization Algorithm
+
+When we use JSON or Toml to represent an initdata metadata, control characters like
+line breaks that do not affect the semantics but will cause the original initdata
+metadata content to change. This fact will make it difficult for us to calculate a
+determined digest from different forms of a same initdata. We need a canonicalization
+algorithm. With this algorithm, all different forms of a same initdata will be calculated
+into a same digest. The difference of forms at least include the following
+
+1. The encoding of the initdata metadata, e.g. JSON and TOML.
+2. Initdata metadata with different expressions but the same semantics, e.g. multiple line
+JSON vs one line JSON like the following
+```json
+{
+  "algorithm": "sha384",
+  "version": "0.1.0",
+  "data": {
+    "key1": "value1",
+    "key2": "value2"
+  },
+  "digest": "0d25829cc97872f0086fbb7eec3a08c0b76899986dc135e9545434ae9eba7d48196efa7323519f00384881b659c02087"
+}
+```
+and
+```json
+{"algorithm":"sha384","version":"0.1.0","data":{"key1":"value1","key2":"value2"},"digest":"0d25829cc97872f0086fbb7eec3a08c0b76899986dc135e9545434ae9eba7d48196efa7323519f00384881b659c02087"}
+```
+
+The algorithm has 2 steps as following
+
+1. Transform the original initdata metadata's `data` field into JSON.
+2. Canonicalize the JSON resulted in step 1. The canonicalization follows [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785). This includes removing extra whitespace characters and ordering of object members.
```

Comment:
Oh. In step 1 we transform it into JSON and do step 2.

If a toml, it will be firstly transform into JSON without changing the meaning.

because TOML standard does not have a canonicalization operation for now.

Let me make it clear in step 1. How about
```
1. Transform the original initdata metadata's `data` field into JSON. This step requires converting
different forms of initdata metadata into JSON without changing the semantics. This is practical
because JSON, TOML and YAML all support key-value, string, and nested formats.
```

Reply:
Oh whoops. I didn't read closely enough. So calculating the digest will always involve turning everything into JSON? This seems slightly cumbersome.

Do we expect the structured init data to change after it is first injected into the guest? It seems like in many cases we could just directly hash the toml file but maybe that isn't a good idea. Not sure.

Reply:
Probably. The core idea is to get a same hash no matter what format, which order of the keys. In JSON we have well-defined canonicalization operation, but for toml and even yaml we do not have one.

So we might need some way to serialize the data into a consistent one. Compared with defining a new algorithm, I prefer to use existing JSON canonicalization

Reply:
if it's just the order of the keys we are concerned about, wouldn't it be sufficient to specify that the hashing will occur in sort-by-keys order?

Reply:
Order is one part, the other is the "control charactors".

This means
```json
{
  "key1": "value1"
}
```
and
```json
{
  "key": "1value1"
}
```

must be different. I wanted to use custom separators between key and value, between key-value pairs. However the existing canonicalization standard helps us to do that without customerized thing.

btw, `serde_json` by default uses B Tree underlying, so its `to_string` could help directly do this.

Reply:
I'm not sure I understand this example. those are two different data structures, why would they be same?

(I'm don't have a strong opinion about the json transformation, but to me the only thing it would achieve is a sorting of keys. that is, unless we allow arbitrary json as payload)

Reply:
Oh. I just want to say that if we do not have good separators, the two examples mentioned might be the same after serialization. Sorry for the confusing example.

I tried to invent some separators. It can be realized, but the cost is complex, as for some separators settings, we can always construct two data with exactly the same serialization results but different semantics based on existing rules.

So instead of inventing a complex segmentation mechanism, it seems that just using the existing JSON canonicalization mechanism will do the trick.

The example I showed is a very simple rule (of course it's inpractical): Concatenate keys and values as input without any separators.

Reply:
duplicated thread with https://github.com/confidential-containers/trustee/pull/348#discussion_r1519396154